### PR TITLE
examples(mentions): convert to plugin and handle keyboard events

### PR DESCRIPTION
**Pull request by ericedem**
#### Is this adding or improving a _feature_ or fixing a _bug_?

Improving an _example_.

#### What's the new behavior?

Update the mentions example to be a plugin and to handle keyboard events.

Here is an illustration of the new keyboard control behavior for the suggestions:

kapture 2019-02-22 at 12 56 21

#### How does this change work?

Here is an overview of the new components added to the example:

```
- MentionPlugin is the core plugin that defines how mentions should be handled
  in slate. It is designed in such a way that you can have multiple instances of
  it in the same editor, like for user mentions and hashtags.

- UserMentionPlugin is the higher level plugin that defines the business logic
  for the mentions implementation like schema restrictions and how the mentions
  themselves should be rendered. It also provides the `<UserMentionSuggestions>`
  component.

- <UserMentionSuggestions> encapsulates business logic around searching for
  suggestions and for handling selection, both with mouse and keybaord.
```

#### Have you checked that...?

* [x] The new code matches the existing patterns and styles.
* [x] The tests pass with `yarn test`.
* [x] The linter passes with `yarn lint`. (Fix errors with `yarn prettier`.)
* [x] The relevant examples still work. (Run examples with `yarn watch`.)

#### Does this fix any issues or need any specific reviewers?

Fixes: #2609

cc: @oyeanuj

### CODE REVIEW

**Comment by oyeanuj** [comment]
@ericedem Apologies that I didn't see this earlier, something with Github notification settings. 

I did a quick parse-thru and this looks great! Thank you for documenting as you went along. I think this is ready to merge for an example.

File: examples/mentions/UserMentionSuggestions.js [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+              {user.username}
+            </Suggestion>
+          )
+        })}
+      </SuggestionList>,
+      root
+    )
+  }
+
+  renderSuggestions() {}
```
Comment (realknack) [inline examples/mentions/UserMentionSuggestions.js]: It looks like this serves no purpose.

**Comment by almaron** [comment]
There's a serious issue calling `setState` from `componentDidUpdate`

**Comment by colinmcd94** [comment]
@ericedem amazing work, thanks for doing this. I'd love to see this get properly merged into master

**Comment by ianstormtaylor** [comment]
Hey @ericedem thank you very much for all of your efforts with Slate. I really appreciate it.

That said, my goal for the examples is for each one to be very self-contained and simple. I don't think it makes sense for us to make the mentions example any more complex (I actually would like to simplify it further). If anyone wants to maintain a full-fledged mentions plugin or example they are more than welcome to. But I can't justify adding the maintenance burden to this repository for example purposes. I hope you understand. I'm sorry I didn't get to this earlier so that I could have saved you some effort.

**Comment by ericedem** [comment]
@ianstormtaylor no worries, I've been on the fence about adding them myself. It wasn't very much work, a lot of this was just copy 🍝 .

**Comment by alainib** [comment]
hello. i don't found MentionPlugin  in doc. 
how to use it please ?

**Pull request closed without merging**